python/pyspark/sql/tests.py

```diff
@@ -4961,6 +4961,31 @@ def foofoo(x, y):
             ).collect
         )
 
+    def test_pandas_udf_detect_unsafe_type_conversion(self):
+        from distutils.version import LooseVersion
+        from pyspark.sql.functions import pandas_udf
+        import pandas as pd
+        import numpy as np
+        import pyarrow as pa
+
+        values = [1.0] * 3
+        pdf = pd.DataFrame({'A': values})
+        df = self.spark.createDataFrame(pdf).repartition(1)
+
+        @pandas_udf(returnType="int")
+        def udf(column):
+            return pd.Series(np.linspace(0, 1, 3))
+
+        udf_boolean = df.select(['A']).withColumn('udf', udf('A'))
+
+        # Since 0.11.0, PyArrow supports the feature to raise an error for unsafe cast.
+        if LooseVersion(pa.__version__) >= LooseVersion("0.11.0"):
```

Comment:
I checked how 0.8.0 is working and it does raise an error for something like overflows, but not truncation like this test. Can you also add a check for overflow not dependent on the version?

Reply:
Ok, sure.

Reply:
BTW, let's bump up the minimal required PyArrow and Pandas version up if possible at 3.0 :-)

Reply:
@BryanCutler Do you mean this same udf raises an error like overflows? I tried with 0.8.0 but it doesn't? Am I missing something?

Reply:
Thanks @BryanCutler. It is clear to me now. I added a test for this.

Reply:
So pyarrow 0.8 - 0.10 raises an error for an integer overflow but not for cast from float to integral types, whereas 0.11 raises an error for both?

Reply:
Seems like the cast from float to integral types are always working without an error in pyarrow < 0.11.

```python
>>> pa.Array.from_pandas(pd.Series([128.0]), type=pa.int8())
<pyarrow.lib.Int8Array object at 0x11c42d940>
[
  -128
]
```

Reply:
Yeah but in pyarrow 0.11.0+ you'd see an error:
```python
>>> import pandas as pd
>>> import pyarrow as pa
>>> pa.__version__
'0.11.1'
>>> pa.Array.from_pandas(pd.Series([128.0]), type=pa.int8())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/array.pxi", line 474, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 169, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 69, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 81, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Floating point value truncated
>>> pa.Array.from_pandas(pd.Series([128.0]), type=pa.int8(), safe=False)
<pyarrow.lib.Int8Array object at 0x7f3ee1a4b868>
[
  -128
]
```

Reply:
Yeah. I was thinking how we should handle this behavior change.
We will have the behavior change anyway regardless of the config, right?
- `safe=True`: we can't use nullable integral types
- `safe=False`: we can't detect an integer overflow

Reply:
I think so. `safe=False` does the type conversion anyway even an overflow.

Reply:
If using `safe=False` by default, it is possible to break user's code without notifications. If using `safe=True` by default, it still breaks user's code, but there is error message so it should let user know what's happen.